// underscore/modules/index.js
// Named Exports
// =============

//     Underscore.js 1.13.6
//     https://underscorejs.org
//     (c) 2009-2022 Jeremy Ashkenas, Julian Gonggrijp, and DocumentCloud and Investigative Reporters & Editors
//     Underscore may be freely distributed under the MIT license.

// Baseline setup.
export { VERSION } from 'underscore/modules/_setup.js';
export { default as restArguments } from 'underscore/modules/restArguments.js';

// Object Functions
// ----------------
// Our most fundamental functions operate on any JavaScript object.
// Most functions in Underscore depend on at least one function in this section.

// A group of functions that check the types of core JavaScript values.
// These are often informally referred to as the "isType" functions.
export { default as isObject } from 'underscore/modules/isObject.js';
export { default as isNull } from 'underscore/modules/isNull.js';
export { default as isUndefined } from 'underscore/modules/isUndefined.js';
export { default as isBoolean } from 'underscore/modules/isBoolean.js';
export { default as isElement } from 'underscore/modules/isElement.js';
export { default as isString } from 'underscore/modules/isString.js';
export { default as isNumber } from 'underscore/modules/isNumber.js';
export { default as isDate } from 'underscore/modules/isDate.js';
export { default as isRegExp } from 'underscore/modules/isRegExp.js';
export { default as isError } from 'underscore/modules/isError.js';
export { default as isSymbol } from 'underscore/modules/isSymbol.js';
export { default as isArrayBuffer } from 'underscore/modules/isArrayBuffer.js';
export { default as isDataView } from 'underscore/modules/isDataView.js';
export { default as isArray } from 'underscore/modules/isArray.js';
export { default as isFunction } from 'underscore/modules/isFunction.js';
export { default as isArguments } from 'underscore/modules/isArguments.js';
export { default as isFinite } from 'underscore/modules/isFinite.js';
export { default as isNaN } from 'underscore/modules/isNaN.js';
export { default as isTypedArray } from 'underscore/modules/isTypedArray.js';
export { default as isEmpty } from 'underscore/modules/isEmpty.js';
export { default as isMatch } from 'underscore/modules/isMatch.js';
export { default as isEqual } from 'underscore/modules/isEqual.js';
export { default as isMap } from 'underscore/modules/isMap.js';
export { default as isWeakMap } from 'underscore/modules/isWeakMap.js';
export { default as isSet } from 'underscore/modules/isSet.js';
export { default as isWeakSet } from 'underscore/modules/isWeakSet.js';

// Functions that treat an object as a dictionary of key-value pairs.
export { default as keys } from 'underscore/modules/keys.js';
export { default as allKeys } from 'underscore/modules/allKeys.js';
export { default as values } from 'underscore/modules/values.js';
export { default as pairs } from 'underscore/modules/pairs.js';
export { default as invert } from 'underscore/modules/invert.js';
export { default as functions,
         default as methods   } from 'underscore/modules/functions.js';
export { default as extend } from 'underscore/modules/extend.js';
export { default as extendOwn,
         default as assign    } from 'underscore/modules/extendOwn.js';
export { default as defaults } from 'underscore/modules/defaults.js';
export { default as create } from 'underscore/modules/create.js';
export { default as clone } from 'underscore/modules/clone.js';
export { default as tap } from 'underscore/modules/tap.js';
export { default as get } from 'underscore/modules/get.js';
export { default as has } from 'underscore/modules/has.js';
export { default as mapObject } from 'underscore/modules/mapObject.js';

// Utility Functions
// -----------------
// A bit of a grab bag: Predicate-generating functions for use with filters and
// loops, string escaping and templating, create random numbers and unique ids,
// and functions that facilitate Underscore's chaining and iteration conventions.
export { default as identity } from 'underscore/modules/identity.js';
export { default as constant } from 'underscore/modules/constant.js';
export { default as noop } from 'underscore/modules/noop.js';
export { default as toPath } from 'underscore/modules/toPath.js';
export { default as property } from 'underscore/modules/property.js';
export { default as propertyOf } from 'underscore/modules/propertyOf.js';
export { default as matcher,
         default as matches } from 'underscore/modules/matcher.js';
export { default as times } from 'underscore/modules/times.js';
export { default as random } from 'underscore/modules/random.js';
export { default as now } from 'underscore/modules/now.js';
export { default as escape } from 'underscore/modules/escape.js';
export { default as unescape } from 'underscore/modules/unescape.js';
export { default as templateSettings } from 'underscore/modules/templateSettings.js';
export { default as template } from 'underscore/modules/template.js';
export { default as result } from 'underscore/modules/result.js';
export { default as uniqueId } from 'underscore/modules/uniqueId.js';
export { default as chain } from 'underscore/modules/chain.js';
export { default as iteratee } from 'underscore/modules/iteratee.js';

// Function (ahem) Functions
// -------------------------
// These functions take a function as an argument and return a new function
// as the result. Also known as higher-order functions.
export { default as partial } from 'underscore/modules/partial.js';
export { default as bind } from 'underscore/modules/bind.js';
export { default as bindAll } from 'underscore/modules/bindAll.js';
export { default as memoize } from 'underscore/modules/memoize.js';
export { default as delay } from 'underscore/modules/delay.js';
export { default as defer } from 'underscore/modules/defer.js';
export { default as throttle } from 'underscore/modules/throttle.js';
export { default as debounce } from 'underscore/modules/debounce.js';
export { default as wrap } from 'underscore/modules/wrap.js';
export { default as negate } from 'underscore/modules/negate.js';
export { default as compose } from 'underscore/modules/compose.js';
export { default as after } from 'underscore/modules/after.js';
export { default as before } from 'underscore/modules/before.js';
export { default as once } from 'underscore/modules/once.js';

// Finders
// -------
// Functions that extract (the position of) a single element from an object
// or array based on some criterion.
export { default as findKey } from 'underscore/modules/findKey.js';
export { default as findIndex } from 'underscore/modules/findIndex.js';
export { default as findLastIndex } from 'underscore/modules/findLastIndex.js';
export { default as sortedIndex } from 'underscore/modules/sortedIndex.js';
export { default as indexOf } from 'underscore/modules/indexOf.js';
export { default as lastIndexOf } from 'underscore/modules/lastIndexOf.js';
export { default as find,
         default as detect } from 'underscore/modules/find.js';
export { default as findWhere } from 'underscore/modules/findWhere.js';

// Collection Functions
// --------------------
// Functions that work on any collection of elements: either an array, or
// an object of key-value pairs.
export { default as each,
         default as forEach } from 'underscore/modules/each.js';
export { default as map,
         default as collect } from 'underscore/modules/map.js';
export { default as reduce,
         default as foldl,
         default as inject } from 'underscore/modules/reduce.js';
export { default as reduceRight,
         default as foldr       } from 'underscore/modules/reduceRight.js';
export { default as filter,
         default as select } from 'underscore/modules/filter.js';
export { default as reject } from 'underscore/modules/reject.js';
export { default as every,
         default as all   } from 'underscore/modules/every.js';
export { default as some,
         default as any  } from 'underscore/modules/some.js';
export { default as contains,
         default as includes,
         default as include  } from 'underscore/modules/contains.js';
export { default as invoke } from 'underscore/modules/invoke.js';
export { default as pluck } from 'underscore/modules/pluck.js';
export { default as where } from 'underscore/modules/where.js';
export { default as max } from 'underscore/modules/max.js';
export { default as min } from 'underscore/modules/min.js';
export { default as shuffle } from 'underscore/modules/shuffle.js';
export { default as sample } from 'underscore/modules/sample.js';
export { default as sortBy } from 'underscore/modules/sortBy.js';
export { default as groupBy } from 'underscore/modules/groupBy.js';
export { default as indexBy } from 'underscore/modules/indexBy.js';
export { default as countBy } from 'underscore/modules/countBy.js';
export { default as partition } from 'underscore/modules/partition.js';
export { default as toArray } from 'underscore/modules/toArray.js';
export { default as size } from 'underscore/modules/size.js';

// `_.pick` and `_.omit` are actually object functions, but we put
// them here in order to create a more natural reading order in the
// monolithic build as they depend on `_.contains`.
export { default as pick } from 'underscore/modules/pick.js';
export { default as omit } from 'underscore/modules/omit.js';

// Array Functions
// ---------------
// Functions that operate on arrays (and array-likes) only, because they’re
// expressed in terms of operations on an ordered list of values.
export { default as first,
         default as head,
         default as take  } from 'underscore/modules/first.js';
export { default as initial } from 'underscore/modules/initial.js';
export { default as last } from 'underscore/modules/last.js';
export { default as rest,
         default as tail,
         default as drop } from 'underscore/modules/rest.js';
export { default as compact } from 'underscore/modules/compact.js';
export { default as flatten } from 'underscore/modules/flatten.js';
export { default as without } from 'underscore/modules/without.js';
export { default as uniq,
         default as unique } from 'underscore/modules/uniq.js';
export { default as union } from 'underscore/modules/union.js';
export { default as intersection } from 'underscore/modules/intersection.js';
export { default as difference } from 'underscore/modules/difference.js';
export { default as unzip,
         default as transpose } from 'underscore/modules/unzip.js';
export { default as zip } from 'underscore/modules/zip.js';
export { default as object } from 'underscore/modules/object.js';
export { default as range } from 'underscore/modules/range.js';
export { default as chunk } from 'underscore/modules/chunk.js';

// OOP
// ---
// These modules support the "object-oriented" calling style. See also
// `underscore.js` and `index-default.js`.
export { default as mixin } from 'underscore/modules/mixin.js';
export { default } from 'underscore/modules/underscore-array-methods.js';
